bevy_lint/src/lints/borrow_of_reborrowable.rs

```diff
@@ -0,0 +1,224 @@
+}
+
+impl<'tcx> LateLintPass<'tcx> for BorrowOfReborrowable {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _: FnKind<'tcx>,
+        decl: &'tcx FnDecl<'tcx>,
+        _: &'tcx Body<'tcx>,
+        _: Span,
+        def_id: LocalDefId,
+    ) {
+        // We are already inside of the function item,
+        // so we can use `instantiate_identity` to discharge the binder
+        let fn_sig = cx.tcx.fn_sig(def_id).instantiate_identity();
+
+        let arg_names = cx.tcx.fn_arg_names(def_id);
+
+        fn_sig.inputs().map_bound(|args| {
+            for (arg_index, arg_ty) in args.iter().enumerate() {
+                let TyKind::Ref(region, ty, Mutability::Mut) = arg_ty.kind() else {
+                    // We only care about `&mut` parameters
+                    continue;
+                };
+
+                let peeled_ty = ty.peel_refs();
+
+                let Some(reborrowable) = Reborrowable::try_from_ty(cx, peeled_ty) else {
+                    // The type is not one of our known re-borrowable types
+                    continue;
+                };
+
+                let is_output_bound_to_arg = fn_sig
+                    .output()
+                    .visit_with(&mut ContainsRegion(*region))
+                    .is_break();
+
+                if is_output_bound_to_arg {
+                    // We don't want to suggest re-borrowing if the return type's
+                    // lifetime is bound to the argument's reference.
+                    // This is because it's impossible to convert something like:
+                    // `for<'a> (&'a mut Commands<'_, '_>) -> EntityCommands<'a>`
+                    // to something like:
+                    // `for<'a> (Commands<'_, '_>) -> EntityCommands<'a>`
+                    // without getting: `error[E0515]: cannot return value referencing function parameter `commands``
+                    continue;
+                }
+
+                let arg_ident = arg_names[arg_index];
+                let span = decl.inputs[arg_index].span.to(arg_ident.span);
+
+                span_lint_and_sugg(
+                    cx,
+                    reborrowable.lint(),
+                    span,
+                    reborrowable.message(),
+                    reborrowable.help(),
+                    reborrowable.suggest(arg_ident, peeled_ty.to_string()),
+                    // Not machine-applicable since the function body may need to
+                    // also be updated to account for the removed ref
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        });
+    }
+}
+
+#[derive(Debug)]
+enum Reborrowable {
+    Commands,
+    // Deferred,
+    // DeferredWorld,
+    EntityCommands,
+    // EntityMut,
+    // EntityMutExcept,
+    // FilteredEntityMut,
+    // FilteredResourcesMut,
+    // Mut,
+    // MutUntyped,
+    NonSendMut,
+    // PtrMut,
+    Query,
+    // QueryIterationCursor,
```

Comment:
What is the reason some of these variants to be commented-out?

Reply:
Whoops I meant to call this out explicitly. I was curious if we wanted to add more lints for these other reborrowable types. We can definitely do all of them, but I wasn't sure if that was overkill (especially for niche types like `QueryIterationCursor`).

Let me know and I can add/remove these!

Reply:
I think I'd enable the extra types. This lint is already pedantic, so people have to opt-in to it, so it's less likely to be annoying. If it becomes a problem people can always create an issue and we can remove it later :)

Reply:
Will do! Should we just merge all the sub-lints into one then? Or should I add new sub-lints to group these new types?

Reply:
Edit: just saw your other comment, I'll combine them haha